fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -709,4 +716,207 @@ public Void visitLogicalTableScan(OptExpression optExpression, Void context) {
+                BinaryPredicateOperator binary = (BinaryPredicateOperator) scalarOperator;
+                ScalarOperator leftChild = scalarOperator.getChild(0);
+                ScalarOperator rightChild = scalarOperator.getChild(1);
+                if (binary.getBinaryType().isEqual()) {
+                    if (leftChild.isColumnRef() && rightChild.isColumnRef()) {
+                        columnEqualityPredicates.add(scalarOperator);
+                    } else if (leftChild.isColumnRef() && rightChild.isConstantRef()) {
+                        rangePredicates.add(scalarOperator);
+                    } else {
+                        residualPredicates.add(scalarOperator);
+                    }
+                } else if (binary.getBinaryType().isRange()) {
+                    if (leftChild.isColumnRef() && rightChild.isConstantRef()) {
+                        rangePredicates.add(scalarOperator);
+                    } else {
+                        residualPredicates.add(scalarOperator);
+                    }
+                }
+            } else {
+                residualPredicates.add(scalarOperator);
+            }
+        }
+        return PredicateSplit.of(Utils.compoundAnd(columnEqualityPredicates), Utils.compoundAnd(rangePredicates),
+                Utils.compoundAnd(residualPredicates));
+    }
+
+    public static ScalarOperator canonizePredicate(ScalarOperator predicate) {
+        if (predicate == null) {
+            return null;
+        }
+        ScalarOperatorRewriter rewrite = new ScalarOperatorRewriter();
+        return rewrite.rewrite(predicate, ScalarOperatorRewriter.DEFAULT_REWRITE_SCAN_PREDICATE_RULES);
+    }
+
+    public static ScalarOperator canonizePredicateForRewrite(ScalarOperator predicate) {
+        if (predicate == null) {
+            return null;
+        }
+        ScalarOperatorRewriter rewrite = new ScalarOperatorRewriter();
+        return rewrite.rewrite(predicate, ScalarOperatorRewriter.MV_SCALAR_REWRITE_RULES);
+    }
+
+    public static boolean isAlwaysFalse(ScalarOperator predicate) {
+        if (predicate instanceof ConstantOperator) {
+            ConstantOperator constant = (ConstantOperator) predicate;
+            if (constant.getType() == Type.BOOLEAN && constant.getBoolean() == false) {
+                return true;
+            }
+        } else if (predicate instanceof CompoundPredicateOperator) {
+            CompoundPredicateOperator compound = (CompoundPredicateOperator) predicate;
+            if (compound.isAnd()) {
+                return isAlwaysFalse(compound.getChild(0)) || isAlwaysFalse(compound.getChild(1));
+            } else if (compound.isOr()) {
+                return isAlwaysFalse(compound.getChild(0)) && isAlwaysFalse(compound.getChild(1));
+            } else if (compound.isNot()) {
+                return isAlwaysTrue(predicate.getChild(0));
+            }
+        }
+        return false;
```

Comment:
why not 
```
ConstantOperator.TRUE.equals(rewrite.rewrite(predicate, ScalarOperatorRewriter.DEFAULT_REWRITE_SCAN_PREDICATE_RULES))?
```

Reply:
ok, I will fix it

Reply:
BTW，I think your code only need check is ConstantOperator.TRUE.equals(predicate), don't need handle compound because you had rewrite